src/magic_folder/test/test_participants.py

```diff
@@ -206,6 +226,252 @@ def test_list(self, collective_contents, rw_collective_dircap):
             ),
         )
 
+    @given(
+        unique_value_dictionaries(
+            author_names(),
+            tahoe_lafs_dir_capabilities(),
+            min_size=2,
+        ),
+        tahoe_lafs_dir_capabilities(),
+    )
+    def test_add(self, collective_contents, rw_collective_dircap):
+        """
+        ``IParticipants.add`` correctly adds a new, previously unknown
+        participant.
+        """
+        # The collective can't be anyone's DMD.
+        assume(rw_collective_dircap not in collective_contents.values())
+        rw_collective_dircap = rw_collective_dircap.encode("ascii")
+
+        # Pick someone in the collective to be us.
+        author = sorted(collective_contents)[0]
+        upload_dircap = collective_contents[author].encode("ascii")
+        upload_dircap_ro = to_readonly_capability(upload_dircap)
+
+        root = create_fake_tahoe_root()
+        http_client = create_tahoe_treq_client(root)
+        tahoe_client = TahoeClient(
+            DecodedURL.from_text(u"http://example.invalid./"),
+            http_client,
+        )
+
+        root._uri.data[rw_collective_dircap] = dumps([
+            u"dirnode",
+            {
+                u"children": {
+                    normalize(author): format_filenode(upload_dircap_ro, {}),
+                },
+            },
+        ])
+
+        root._uri.data[upload_dircap] = dumps([
+            u"dirnode",
+            {u"children": {}},
+        ])
+
+        participants = participants_from_collective(
+            rw_collective_dircap,
+            upload_dircap,
+            tahoe_client,
+        )
+
+        # add all the "other" participants using .add() API
+        for name, dircap in collective_contents.items():
+            if name == author:
+                continue
+            participants.add(
+                RemoteAuthor(name, VerifyKey(os.urandom(32))),
+                to_readonly_capability(dircap),
+            )
+
+        # confirm we added all the right participants by using the
+        # list() API
```

Comment:
If I understand the suggestion properly, I think the answer is "no" --  because that's what the `Participants` API is supposed to be doing -- that is, abstracting away "how Tahoe represents things" .. 

That said, we're already somewhat tied to that representation because a few lines up we have to put fake Tahoe data in which means we're pretty commited to the JSON returned by Tahoe's WebUI (and really all we'd be testing there is that our Fake is correct still, which arguably should  be in other tests).

So, I think the Tahoe fake (with its tests) should handle checking that the shape of the JSON is good after e.g. calling "`TahoeClient.add_entry_to_directory()`" and this code/test can just look at the output of the higher-level API (`IParticipants.list()`)

Reply:
> If I understand the suggestion properly, I think the answer is "no" -- because that's what the Participants API is supposed to be doing -- that is, abstracting away "how Tahoe represents things" ..

Well, the thing we aren't testing is if we are representing the abstraction "reasonably" or "as we expect". Which, granted, depends somewhat on our fake being accurate. Given that we want to maintain forward and backwards compatibility, at some point, we probably do want tests that make assertions about what calling our abstractions does on the underlying tahoe level.

That said, this may not be the right place for those tests, and they should instead by in the integration tests, where we inspect the state in a real tahoe cluster after our operations.